src/tools/cephfs/cephfs_shell.py

```diff
@@ -0,0 +1,844 @@
+
+def is_file_exists(file_name, dir = ''):
+    if dir == '':
+        dir = cephfs.getcwd()
+    elif not isinstance(dir, bytes):
+        dir = to_bytes(dir)
+    if not isinstance(file_name, bytes):
+        file_name = to_bytes(file_name)
+    return len([i for i in set(list_items(dir)) if i.d_name == file_name and not i.is_dir()]) > 0
+
+def print_list(words, termwidth = 79):
+    if not words :
+        return
+    width = max([word_len(word) for word in words]) + 2
+    nwords = len(words)
+    ncols = max(1, (termwidth + 1) // (width + 1))
+    nrows = (nwords + ncols - 1) // ncols
+    for row in range(nrows):
+        for i in range(row, nwords, nrows):
+            word = words[i]
+            if word[0] == '\x1b':
+                print('%-*s'% (width + 9, words[i]), end = '\n'if i + nrows >= nwords else '')
+            else:
+                print('%-*s'% (width, words[i]), end = '\n'if i + nrows >= nwords else '')
+
+def copy_from_local(local_path, remote_path):
+    stdin = -1
+    if local_path == '-':
+        data = ''.join([line for line in sys.stdin])
+        file_size = len(data)
+    else :
+        file_ = open(local_path, 'rb')
+        stdin = 1
+        file_size = os.path.getsize(local_path)
+        print('File {}: {}'.format(local_path, file_size))
+    if is_file_exists(remote_path):
+        return
+    fd = cephfs.open(to_bytes(remote_path), 'w', 0o666) 
+    if file_size == 0:
+        return  
+    progress = 0
+    while True:
+        data = file_.read(65536)
+        if not data:
+            break
+        wrote = cephfs.write(fd, data, progress) 
+        if wrote < 0:
+            break
+        progress += wrote
+        print('{} of {} bytes read ({}%)'.format(progress, file_size, int(progress / file_size * 100)), end = '\r')
+    cephfs.close(fd)
+    if stdin > 0 : 
+        file_.close()
+    print()
+
+def copy_to_local(remote_path, local_path):
+    print('lp', local_path)
+    if not os.path.exists(local_path.rsplit('/', 1)[0]) :
+          os.makedirs(local_path.rsplit('/', 1)[0], exist_ok = True)
+    fd = None
+    if len(remote_path.rsplit('/', 1)) > 2 and remote_path.rsplit('/', 1)[1] == '':
+        return
+    if local_path != '-':
+        fd = open(local_path, 'wb+')
+    file_ = cephfs.open(to_bytes(remote_path), 'r')
+    file_size = cephfs.stat(remote_path).st_size
+    if file_size <= 0 :
+        return
+    print('File {}: {}'.format(remote_path, file_size))
+    progress = 0
+    for chunk_start, chunk_size in get_chunks(file_size):
+        file_chunk = cephfs.read(file_, chunk_start, chunk_size)
+        progress += len(file_chunk)
+        if fd :
+            fd.write(file_chunk)
+            print('{} of {} bytes read ({}%)'.format(progress, file_size, int(progress / file_size * 100)), end = '\r') 
+        else :
+            print(file_chunk.decode('utf-8'))  
+    print() 
+    cephfs.close(file_)
+    if fd : 
+        fd.close()   
+
+def dirwalk(dir_name, giveDirs=0):
```

Comment:
Is `giveDirs` supposed to only return sub-directories?

Reply:
it is for returning empty directories.

Reply:
In which case do you not want to return empty directories?

Reply:
When we are copying a directory tree, we even have to copy the directory which is empty right?

Reply:
Right, but I'm asking when do we **not** want to return empty directories?

Reply:
No there is no such case.